### PR TITLE
servenv: Allow for explicit bind address

### PR DESCRIPTION
**Pull request by dbussink**
In environments where there's potentially multiple address available, it can be useful to only bind on a specific address. This avoids making the process available on any other interfaces for a better security model.

We also make use of this in the CI tests to bind to the local network address to avoid exposing any CI processes to any external address that might be available.

Docs PR: https://github.com/vitessio/website/pull/1600

## Related Issue(s)

Fixes https://github.com/vitessio/vitess/issues/13891

## Checklist

-   [x] "Backport to:" labels have been added if this change should be back-ported
-   [x] Tests were added or are not required
-   [x] Did the new or modified tests pass consistently locally and on the CI
-   [x] Documentation was added or is not required

### CODE REVIEW

File: go/vt/servenv/run.go [hunk 1]
```diff
@@ -37,14 +37,14 @@ var (
...
-func Run(port int) {
+func Run(bindAddress string, port int) {
 	populateListeningURL(int32(port))
 	createGRPCServer()
 	onRunHooks.Fire()
 	serveGRPC()
 	serveSocketFile()
 
-	l, err := net.Listen("tcp", fmt.Sprintf(":%v", port))
+	l, err := net.Listen("tcp", fmt.Sprintf("%s:%v", bindAddress, port))
```
Comment (ajm188) [inline go/vt/servenv/run.go:47]: we should be using `net.JoinHostPort` here instead (https://pkg.go.dev/net#JoinHostPort)

File: go/flags/endtoend/mysqlctld.txt [hunk 2]
```diff
@@ -22,6 +22,7 @@ Flags:
       --alsologtostderr                                                  log to standard error as well as files
       --app_idle_timeout duration                                        Idle timeout for app connections (default 1m0s)
       --app_pool_size int                                                Size of the connection pool for app connections (default 40)
+      --bind_address string                                              Bind address for the server.
```
Comment (ajm188) [inline go/flags/endtoend/mysqlctld.txt:25]: you'll need to generate the website docs for all these (`make generate-docs` in the website dir, see also the envvars you'll need to set)

**Review by ajm188 — APPROVED** [review]

File: go/vt/servenv/run.go (hunk 1 above)
Comment (frouioui) [reply go/vt/servenv/run.go]: Fixed :)

File: go/flags/endtoend/mysqlctld.txt (hunk 2 above)
Comment (frouioui) [reply go/flags/endtoend/mysqlctld.txt:25]: Got it! Thank you for the pointer @ajm188

Comment (frouioui) [reply go/flags/endtoend/mysqlctld.txt:25]: https://github.com/vitessio/website/pull/1600

File: go/test/endtoend/cluster/vtctld_process.go [hunk 3]
```diff
@@ -66,6 +66,14 @@ func (vtctld *VtctldProcess) Setup(cell string, extraArgs ...string) (err error)
 		"--port", fmt.Sprintf("%d", vtctld.Port),
 		"--grpc_port", fmt.Sprintf("%d", vtctld.GrpcPort),
 	)
+
+	if v, err := GetMajorVersion("vtctld"); err != nil {
+		return err
+	} else if v >= 18 {
+		vtctld.proc.Args = append(vtctld.proc.Args, "--bind_address", "127.0.0.1")
+		vtctld.proc.Args = append(vtctld.proc.Args, "--grpc_bind_address", "127.0.0.1")
```
Comment (deepthi) [inline go/test/endtoend/cluster/vtctld_process.go:74]: New flags should use dash and not underscore as a separator.

File: go/vt/servenv/servenv.go [hunk 4]
```diff
@@ -266,6 +267,7 @@ func FireRunHooks() {
 func RegisterDefaultFlags() {
 	OnParse(func(fs *pflag.FlagSet) {
 		fs.IntVar(&port, "port", port, "port for the server")
+		fs.StringVar(&bindAddress, "bind_address", bindAddress, "Bind address for the server.")
```
Comment (deepthi) [inline go/vt/servenv/servenv.go:270]: 1. Why is the flag called `bind_address` (or `bind-address`)? Why not just `host`?
2. What happens if it is not specified? That should be included in the flag help text.

Comment (frouioui) [reply go/vt/servenv/servenv.go:270]: 1. I think it means the same thing, I don't mind changing it to host. I re-used what was originally on the PR.
2. The output from viper includes the default value for this flag (127.0.0.1)

File: go/test/endtoend/cluster/vtctld_process.go (hunk 3 above)
Comment (frouioui) [reply go/test/endtoend/cluster/vtctld_process.go]: Fixed via 4fd296bc2b

File: go/vt/servenv/grpc_server.go [hunk 5]
```diff
@@ -127,7 +127,7 @@ var (
 func RegisterGRPCServerFlags() {
 	OnParse(func(fs *pflag.FlagSet) {
 		fs.IntVar(&gRPCPort, "grpc_port", gRPCPort, "Port to listen on for gRPC calls. If zero, do not listen.")
-		fs.StringVar(&gRPCBindAddress, "grpc_bind_address", gRPCBindAddress, "Bind address for gRPC calls. If empty, listen on all addresses.")
+		fs.StringVar(&gRPCBindAddress, "grpc_bind-address", gRPCBindAddress, "Bind address for gRPC calls. If empty, listen on all addresses.")
```
Comment (ajm188) [inline go/vt/servenv/grpc_server.go:130]: should we go full-dash here? (i know the other flags use `grpc_` as a prefix, we can also "normalize": https://github.com/spf13/pflag#mutating-or-normalizing-flag-names)

Comment (frouioui) [reply go/vt/servenv/grpc_server.go]: I reverted it to use only `_`, we can find a longer term fix for all flags later.

File: go/vt/servenv/servenv.go (hunk 4 above)
Comment (frouioui) [reply go/vt/servenv/servenv.go]: Apologies, the default value was indeed missing, I added a text to explain that if the flag is left empty, all addresses will be used.

File: go/vt/servenv/grpc_server.go [hunk 6]
```diff
@@ -124,6 +127,7 @@ var (
 func RegisterGRPCServerFlags() {
 	OnParse(func(fs *pflag.FlagSet) {
 		fs.IntVar(&gRPCPort, "grpc_port", gRPCPort, "Port to listen on for gRPC calls. If zero, do not listen.")
+		fs.StringVar(&gRPCBindAddress, "grpc_bind_address", gRPCBindAddress, "Bind address for gRPC calls. If empty, listen on all addresses.")
```
Comment (harshit-gangal) [inline go/vt/servenv/grpc_server.go:130]: new flags should be with `-`

**Review by harshit-gangal — APPROVED** [review]

**Review by GuptaManan100 — APPROVED** [review]
LGTM!

**Pull request merged**